src/ls_server.rs

```diff
@@ -376,6 +406,14 @@ impl MessageReader for StdioMsgReader {
         let mut buffer = String::new();
         handle_err!(io::stdin().read_line(&mut buffer), "Could not read from stdin");
 
+        if buffer.len() == 0 && ls_service.shut_down.load(Ordering::SeqCst) {
+            self.logger.log("Expected and got empty header when shutting down\n");
+            return None;
+        } else if ls_service.shut_down.load(Ordering::SeqCst) {
+            self.logger.log("Got non-empty header when shutting down\n");
+            return None;
+        }
+
```

Comment:
Rather than take an LsService here (which seems like unnecessary coupling), could you handle this outside the read_message call?